pkg/wireguard/agent/node_handler.go

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package agent
+
+import (
+	"context"
+	"net"
+
+	"github.com/cilium/cilium/api/v1/models"
+	datapath "github.com/cilium/cilium/pkg/datapath/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+)
+
+// NodeAdd is called when a node is discovered for the first time.
+func (a *Agent) NodeAdd(newNode nodeTypes.Node) error {
+	if newNode.IsLocal() {
+		return nil
+	}
+
+	if newNode.WireguardPubKey != "" {
+		newIP4 := newNode.GetNodeIP(false)
+		newIP6 := newNode.GetNodeIP(true)
+
+		if err := a.UpdatePeer(newNode.Fullname(), newNode.WireguardPubKey, newIP4, newIP6); err != nil {
+			log.WithError(err).
+				WithField(logfields.NodeName, newNode.Fullname()).
+				Warning("Failed to update wireguard configuration for peer")
+		}
+	}
+
+	// FIXME error or nil?
+	return nil
+}
+
+// NmdeUpdate is called when a node definition changes. Both the old
+// and new node definition is provided. NodeUpdate() is never called
+// before NodeAdd() is called for a particular node.
+func (a *Agent) NodeUpdate(oldNode, newNode nodeTypes.Node) error {
+	if newNode.IsLocal() {
```

Comment:
```suggestion
	if newNode.IsLocal() || newNode.WireguardPubKey == "" {
```

And then drop the if statement below